tests/benchmarks/data_structure_perf/rbtree_perf/src/rbtree_perf.c

```diff
@@ -0,0 +1,159 @@
+struct container_node {
+	struct rbnode node;
+	int value;
+};
+static struct rbnode nodes[TREE_SIZE];
+static struct rbtree tree;
+
+/* Our "lessthan" is just the location of the struct */
+bool node_lessthan(struct rbnode *a, struct rbnode *b)
+{
+	return a < b;
+}
+
+/**
+ * @brief Test whether rbtree node struct is embedded
+ * in any user struct
+ *
+ * @details Initialize an user defined structure contains
+ * rbtree node.And into a rbtree and enumerate
+ * the rbtree nodes.
+ *
+ * verify that the value enumerated is equal to the
+ * value initialized.If the verification pass,the user
+ * defined structure works.
+ * verify "for each" style APIs work.
+ *
+ * @ingroup lib_rbtree_tests
+ *
+ * @see RB_FOR_EACH(),RB_FOR_EACH_CONTAINER()
+ */
+void test_rbtree_container(void)
+{
+	int count = 0;
+	static struct rbtree test_tree_l;
+	static struct container_node tree_node[10];
+	static struct container_node *c_foreach_node;
+	static struct rbnode *foreach_node;
+
+	test_tree_l.lessthan_fn = node_lessthan;
+	for (unsigned int i = 0; i < ARRAY_SIZE(tree_node); i++) {
+		tree_node[i].value = i;
+		rb_insert(&test_tree_l, &tree_node[i].node);
+	}
+
+	RB_FOR_EACH(&test_tree_l, foreach_node) {
+		zassert_true(CONTAINER_OF(foreach_node, struct container_node,
+		node)->value == count, "RB_FOR_EACH failed");
+		count++;
+	}
+
+	count = 0;
+
+	RB_FOR_EACH_CONTAINER(&test_tree_l, c_foreach_node, node) {
+		zassert_true(c_foreach_node->value == count,
+		"RB_FOR_EACH_CONTAINER failed");
+		count++;
+	}
+}
+
+/* initalize and insert a tree */
+void init_tree(struct rbtree *tree, int size)
+{
+	tree->lessthan_fn = node_lessthan;
+
+	for (int i = 0; i < size; i++) {
+		rb_insert(tree, &nodes[i]);
+	}
+}
+
+uint32_t node_height;
+int search_height_recurse(struct rbnode *node, struct rbnode *finial_node)
+{
+	int side = 0;
+
+	if (node == finial_node) {
+		return 0;
+	}
+	if (node != NULL && finial_node != NULL) {
+		side = tree.lessthan_fn(finial_node, node) ? 0 : 1;
+		struct rbnode *ch = z_rb_child(node, side);
+
+		node_height++;
+		search_height_recurse(ch, finial_node);
+	}
+
+	return 0;
+}
+
+void verify_rbtree_perf(struct rbnode *root, struct rbnode *test)
+{
+	/* zephyr can't do floating-point arithmetic,
+	 * so manual:dlog_N = 2log(TREE_SIZE) = 18
+	 */
+	int dlog_N = 18;
+
+	node_height = 0;
+	search_height_recurse(root, test);
+	zassert_true(node_height <= dlog_N, NULL);
+}
+
+/**
+ * @brief Test some operations of rbtree are running in
+ * logarithmic time
+ *
+ * @details
+ * Defining a tree, searching height of maxmum node, minmum node
+ * and the node inserted or removed.And record those heights.
+ *
+ * verify that search heights are less than the height of
+ * worset-case condition(<= 2lg(N)).(N means the size of tree)
+ *
+ * @ingroup lib_rbtree_tests
+ *
+ * @see rb_get_min(),rb_get_max()
+ */
+void test_rbtree_perf(void)
+{
+	memset(&tree, 0, sizeof(tree));
+	memset(nodes, 0, sizeof(nodes));
```

Comment:
nitpick, why are you initializing them ? Aren't they static global variables not used before ?

Reply:
agreed these are in BSS should already be zeroed

Reply:
Good to know, I have removed 'memset'.